did_core/did_parser_nom/src/did/mod.rs

```diff
@@ -0,0 +1,127 @@
+pub(crate) mod parsing;
+
+use std::{
+    convert::TryFrom,
+    fmt::{Display, Formatter},
+    str::FromStr,
+};
+
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use self::parsing::parse_did;
+use crate::{error::ParseError, DidRange, DidUrl};
+
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct Did {
+    did: String,
+    method: Option<DidRange>,
+    namespace: Option<DidRange>,
+    id: DidRange,
```

Comment:
Hmm, I am not super sure about this, but is it still meaningful to use ranges for internal representation? The output of the high level parsers such as `parse_qualified_sovrin_did` is on happy path tuple of the individual fragments `(&'a str, &'a str, Option<&'a str>, &'a str)` - each being slice of the `did` attribute. Why not to just store `&str` slices as references to `did: String`? (I might be overseeing some challenges this might imply.)

In the original implementation, these Ranges were inherently created as our implementation went over the string, but since we have nom now, it seems we have to go extra mile to construct ranges after parsing is done. After `struct Did` is constructed, we really utilize the ranges somewhat special - seems that nom-produced slices would do?

Reply:
No, the main reason the ranges were conceived in first place was to avoid exactly this - dealing with self-referential structs.

Reply:
I see fair, yeah now I see the self-ref struct I am suggesting gets tricky real quick, even if willing to give up on mutability of the struct.